subsys/usb/host/usbh_class.c

```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright (c) 2025 Nordic Semiconductor ASA
+ * Copyright 2025 NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/usb/usbh.h>
+#include <zephyr/logging/log.h>
+
+#include "usbh_class.h"
+#include "usbh_class_api.h"
+#include "usbh_desc.h"
+#include "usbh_host.h"
+
+LOG_MODULE_REGISTER(usbh_class, CONFIG_USBH_LOG_LEVEL);
+
+void usbh_class_init_all(struct usbh_context *const uhs_ctx)
+{
+	int ret;
+
+	usbh_host_lock(uhs_ctx);
+
+	STRUCT_SECTION_FOREACH(usbh_class_node, c_node) {
+		struct usbh_class_data *const c_data = c_node->c_data;
+
+		if (c_node->state != USBH_CLASS_STATE_IDLE) {
+			LOG_DBG("Skipping '%s' in state %u", c_data->name, c_node->state);
+			continue;
+		}
+
+		ret = usbh_class_init(c_data, uhs_ctx);
+		if (ret != 0) {
+			LOG_WRN("Failed to initialize class %s (%d)", c_data->name, ret);
+			c_node->state = USBH_CLASS_STATE_ERROR;
+		}
+	}
+
+	usbh_host_unlock(uhs_ctx);
+}
+
+void usbh_class_remove_all(struct usb_device *const udev)
+{
+	int ret;
+
+	k_mutex_lock(&udev->mutex, K_FOREVER);
+
+	STRUCT_SECTION_FOREACH(usbh_class_node, c_node) {
+		struct usbh_class_data *const c_data = c_node->c_data;
+
+		if (c_data->udev == udev) {
+			ret = usbh_class_removed(c_data);
+			if (ret != 0) {
+				LOG_ERR("Failed to handle device removal for each class (%d)", ret);
+				c_node->state = USBH_CLASS_STATE_ERROR;
+				continue;
+			}
+
+			/* The class instance is now free to bind to a new device */
+			c_data->udev = NULL;
+			c_node->state = USBH_CLASS_STATE_IDLE;
```

Comment:
I think it is not the right way to manage resources and likely will be difficult to rework later.